File: Baekjoon/Silver/S_1012/S2_1012_유기농_배추.py
import sys
sys.setrecursionlimit(10000)

# DFS로 인접한 배추들 방문
def dfs(x, y, field, visited, M, N):
    # 상, 하, 좌, 우 방향
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    visited[y][x] = True

    for dx, dy in directions:
        nx, ny = x + dx, y + dy
        if 0 <= nx < M and 0 <= ny < N and not visited[ny][nx] and field[ny][nx] == 1:
            dfs(nx, ny, field, visited, M, N)


# 테스트 케이스 개수 입력
T = int(input())

for _ in range(T):
    # 테스트 케이스 입력 후 분리
    M, N, K = map(int, input().split())
    field = [[0] * M for _ in range(N)] # 배추밭
    visited = [[False] * M for _ in range(N)] # 방문 여부

    for _ in range(K):
        # 배추 위치 입력 후 분리
        x, y = map(int, input().split())
        field[y][x] = 1

    worm_count = 0 # 배추흰지렁이 마리 수

    for y in range(N):
        for x in range(M):
            if field[y][x] == 1 and not visited[y][x]:
                # 새로운 배추 묶음이 있으면 지렁이 수 증가
                dfs(x, y, field, visited, M, N)
                worm_count += 1
    
    print(worm_count)

File: Baekjoon/Silver/S_1012/S2_1012_유기농_배추.java
package Baekjoon.Silver.S_1012;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class S2_1012_유기농_배추 {
    static int T, M, N, K;
    static StringBuilder strBuild = new StringBuilder();
    static int[] dx = {1, -1, 0, 0}; // x: 좌우
    static int[] dy = {0, 0, 1, -1}; // y: 상하
    static int[][] field;
    static boolean[][] visited;
    static int wormCount;

    public static void main(String[] args) throws IOException {
        BufferedReader bufRead = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer strToken = null;

        T = Integer.parseInt(bufRead.readLine());

        // 테스트 케이스 만큼 반복
        for (int t = 0; t < T; t++) {
            strToken = new StringTokenizer(bufRead.readLine());
            M = Integer.parseInt(strToken.nextToken()); // 가로 길이
            N = Integer.parseInt(strToken.nextToken()); // 세로 길이
            K = Integer.parseInt(strToken.nextToken()); // 배추 위치 개수

            field = new int[N][M];
            visited = new boolean[N][M];

            // 배추 위치 개수 만큼 반복
            for (int i = 0; i < K; i++) {
                strToken = new StringTokenizer(bufRead.readLine());
                int X = Integer.parseInt(strToken.nextToken()); // 열
                int Y = Integer.parseInt(strToken.nextToken()); // 행
                field[Y][X] = 1;
            }

            wormCount = 0;

            // 전체 밭 탐색
            for (int y = 0; y < N; y++) {
                for (int x = 0; x < M; x++) {
                    if (field[y][x] == 1 && !visited[y][x]) {
                        dfs(y, x);
                        wormCount++;
                    }
                }
            } 
            strBuild.append(wormCount).append('\n');
        }
        System.out.println(strBuild.toString());
    }

    // DFS로 배추가 연결된 모든 위치 방문 처리
    static void dfs(int y, int x) {
        visited[y][x] = true;

        for (int dir = 0; dir < 4; dir++) {
            int ny = y + dy[dir];
            int nx = x + dx[dir];

            if (ny < 0 || ny >= N || nx < 0 || nx >= M) continue;

            if (field[ny][nx] == 1 && !visited[ny][nx]) {
                dfs(ny, nx);
            }
        }
    }
}
